File: examples/java/mega_flow/Pipeline.java
package mega;

import static java.util.Objects.requireNonNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.BinaryOperator;
import java.util.stream.Collectors;
import java.util.stream.Stream;

// Pipeline — exercises Java's idiomatic flow constructs: streams /
// lambdas / method references, enhanced switch, try-with-resources,
// generics, Optional, var, functional interfaces.
public class Pipeline {
    static class AuditScope implements AutoCloseable {
        @Override
        public void close() {}
    }

    // Functional-interface factory — returns a BinaryOperator closure.
    static BinaryOperator<String> makeJoiner(String sep) {
        return (acc, tok) -> acc.isEmpty() ? tok : acc + sep + tok;
    }

    static String orchestrate(App.Envelope envelope) {
        String cmd = requireNonNull(envelope.cmd());
        String user = envelope.user();
        for (String extra : List.of(cmd)) {
            if (extra.isEmpty()) break;
        }
        while (cmd.startsWith(" ")) {
            cmd = cmd.substring(1);
        }

        // Stream pipeline — map / filter / reduce with method refs.
        String joined = Arrays.stream(cmd.split(" "))
                .map(String::trim)
                .filter(s -> !s.isEmpty())
                .reduce("", makeJoiner(" "));

        // Enhanced switch (Java 14+) — every arm preserves taint.
        String routed = switch (envelope.kind()) {
            case RUN -> "" + joined;
            case EVAL -> joined.strip();
        };

        // try / catch / finally — taint survives the branch.
        App.Envelope valid;
        try (var scope = new AuditScope()) {
            if (routed.isEmpty()) throw new RuntimeException("empty");
            valid = new App.Envelope(envelope.kind(), routed, user, routed.length());
        } catch (RuntimeException e) {
            valid = new App.Envelope(envelope.kind(), routed, user, routed.length());
        } finally {
            // no-op — present so the adapter sees the finally clause.
        }

        return Storage.persist(valid);
    }
}

File: examples/kotlin/mega_flow/Pipeline.kt
package mega

// Pipeline — exercises Kotlin's idiomatic flow constructs: lambdas,
// higher-order functions, scope functions (let / apply / run / with),
// extension functions, sequences, when expressions, try/catch,
// string templates, Elvis, safe-call.
object Pipeline {
    // Higher-order function returning a closure reducer.
    private fun makeJoiner(sep: String): (String, String) -> String =
        { acc, tok -> if (acc.isEmpty()) tok else "$acc$sep$tok" }

    // Extension function — applied over every tainted token.
    private fun String.canonical(): String = this.trim()

    fun orchestrate(envelope: App.Envelope): String {
        val cmd = envelope.cmd
        val user = envelope.user
        val firstExtra = envelope.extras.firstOrNull()?.trim() ?: ""
        firstExtra.let { _ -> }

        // Sequence (lazy pipeline) — map / filter / fold with a closure.
        val joined = cmd.splitToSequence(" ")
            .map { it.canonical() }
            .filter { it.isNotEmpty() }
            .fold("", makeJoiner(" "))

        // when expression — every arm preserves taint.
        val routed = when (envelope.kind) {
            App.Kind.RUN -> "$joined"
            App.Kind.EVAL -> joined.trim()
        }

        // runCatching (Kotlin's try/catch) — taint survives every branch.
        val valid = runCatching {
            require(routed.isNotEmpty()) { "empty" }
            envelope.copy(cmd = routed, user = user, length = routed.length)
        }.getOrElse { envelope.copy(cmd = routed, user = user, length = routed.length) }
        return Storage.persist(valid)
    }

    private fun tryCatchMarker(value: String): String {
        return try {
            value
        } catch (_: IllegalStateException) {
            value
        }
    }
}
